### PR TITLE
ECS Exec in a running container

### PR DESCRIPTION
**Pull request by gsweetwood**
## Problem
Desire to ECS exec commands into running containers.

## Solution
Add menu item and command to container nodes and allow the user to input a command to run.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by gsweetwood** [comment]
This will also show a duplicate of the changes made in  #1847 until that PR is merged.  This PR is functional but still in draft, comments are always welcome.

File: src/ecs/commands/runCommandInContainer.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+import { EcsContainerNode } from '../explorer/ecsContainerNode'
+import { recordEcsRunExecuteCommand } from '../../shared/telemetry/telemetry.gen'
+import { DefaultSettingsConfiguration } from '../../shared/settingsConfiguration'
+import { extensionSettingsPrefix } from '../../shared/constants'
+
+export async function runCommandInContainer(node: EcsContainerNode, window = Window.vscode()): Promise<void> {
+    getLogger().debug('RunCommandInContainer called for: %O', node)
+
+    const verifySSMPluginResponse = await new ChildProcess(true, 'session-manager-plugin').run()
+    if (!verifySSMPluginResponse.stdout.startsWith('The Session Manager plugin was installed successfully')) {
```
Comment (justinmk3) [inline src/ecs/commands/runCommandInContainer.ts:22]: Dunno if we can rely on this output message. It should be enough to check that the exitCode === 0.

File: src/ecs/commands/runCommandInContainer.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        task,
+        '--container',
+        node.continerName,
+        '--command',
+        command,
+        '--interactive'
+    ).run()
+
+    getLogger().info(response.stdout)
+    if (response.exitCode !== 0 || response.stderr) {
```
Comment (justinmk3) [inline src/ecs/commands/runCommandInContainer.ts:128]: presense of stderr is not an error, if exitCode is zero. Unless there is a special case for `aws` or `aws ecs execute-command`, that you are aware of (in that case, add a 1-line comment specifying this unusual case)

File: src/shared/clients/ecsClient.ts [hunk 3]
```diff
@@ -77,6 +77,19 @@ export class DefaultEcsClient {
...
 
+    public async listTasks(cluster: string, serviceName: string): Promise<string[]> {
+        const sdkClient = await this.createSdkClient()
+
+        const params: ECS.ListTasksRequest = { cluster: cluster, serviceName: serviceName, desiredStatus: 'RUNNING' }
+        try {
+            const listTasksResponse = await sdkClient.listTasks(params).promise()
+            return listTasksResponse.taskArns ?? []
+        } catch (error) {
+            getLogger().error(`ecs: Failed to get tasks for Cluster ${cluster} and Service ${serviceName}: `)
```
Comment (justinmk3) [inline src/shared/clients/ecsClient.ts:88]: variable values should go in quotes, if they must be in the middle of the message. This is mentioned in [guidelines](https://github.com/aws/aws-toolkit-vscode/blob/master/docs/CODE_GUIDELINES.md) (and previous PRs).
```suggestion
            getLogger().error(`ecs: Failed to get tasks for Cluster "${cluster}" and Service "${serviceName}": `)
```

alternatively you could use this feature of console.log / Winston:

            getLogger().error(`ecs: Failed to get tasks for Cluster %O and Service %O: %O`, cluster, serviceName, error)

BTW, it is not really valid to assume that the `throw error` will sink to the log, so typically we log the error explicitly (thus `error` is given in the sample above).

File: src/ecs/commands/runCommandInContainer.ts (hunk 1 above)
Comment (gsweetwood) [reply src/ecs/commands/runCommandInContainer.ts:22]: done

File: src/ecs/commands/runCommandInContainer.ts (hunk 2 above)
Comment (gsweetwood) [reply src/ecs/commands/runCommandInContainer.ts:128]: no special case. corrected.

File: package.nls.json [hunk 4]
```diff
@@ -166,6 +167,14 @@
     "AWS.command.ecr.createRepository": "Create Repository...",
     "AWS.command.ecr.deleteRepository": "Delete Repository...",
     "AWS.command.ecr.deleteTag": "Delete Tag...",
+    "AWS.command.runCommandInContainer": "Run command in container",
+    "AWS.command.ecs.runCommandInContainer.warnBeforeExecute": "Command may modify the running container {0}. Are you sure?",
+    "AWS.command.ecs.runCommandInContainer.yesDoNotShowAgain": "Yes, and don't ask again",
```
Comment (bryceitoc9) [inline package.nls.json:172]: Nit: this matches `AWS.message.prompt.defaultRegionHidden.alwaysAdd`; can you make this generic and combine the strings?

File: package.nls.json [hunk 5]
```diff
@@ -166,6 +167,14 @@
...
+    "AWS.command.ecs.runCommandInContainer.noCliFound": "Please install the AWS CLI before proceding",
+    "AWS.command.ecs.runCommandInContainer.noPluginFound": "Please install the SSM Session Manager plugin before proceding",
+    "AWS.command.ecs.runCommandInContainer.noTasks": "There are no running tasks for the service: {0}",
+    "AWS.command.ecs.runCommandInContainer.prompt": "Enter the command to run in container",
```
Comment (bryceitoc9) [inline package.nls.json:176]: Nit: "Enter the command to run in container {0}" (with the param being the container's name)?

File: src/ecs/commands/runCommandInContainer.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    const verifySsmPluginResponse = await new ChildProcess(true, 'session-manager-plugin').run()
+    if (verifySsmPluginResponse.exitCode !== 0) {
+        window.showErrorMessage(
+            localize(
+                'AWS.command.ecs.runCommandInContainer.noPluginFound',
+                'Please install the SSM Session Manager plugin before proceding'
+            )
+        )
+        throw new Error('The SSM Session Manager plugin is not installed.')
+    }
```
Comment (bryceitoc9) [inline src/ecs/commands/runCommandInContainer.ts:38]: Nit: Can you combine these into a single function? You can also combine the string.

Nit: replace "AWS" in the localized string with `getIdeProperties().company`

File: src/ecs/commands/runCommandInContainer.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+                'There are no running tasks for the service: {0}',
+                node.serviceName
+            )
+        )
+        return
+    }
+
+    const quickPick = picker.createQuickPick({
+        options: {
+            title: 'Choose a task',
```
Comment (bryceitoc9) [inline src/ecs/commands/runCommandInContainer.ts:58]: Localize

File: src/ecs/commands/runCommandInContainer.ts [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+            ),
+            { modal: true },
+            localize('AWS.generic.response.yes', 'Yes'),
+            localize('AWS.command.ecs.runCommandInContainer.yesDoNotShowAgain', "Yes, and don't ask again")
+        )
+        if (choice === undefined) {
+            getLogger().debug('ecs: Cancelled runCommandInContainer')
+            return
+        } else if (
+            choice === localize('AWS.command.ecs.runCommandInContainer.yesDoNotShowAgain', "Yes, and don't ask again")
```
Comment (bryceitoc9) [inline src/ecs/commands/runCommandInContainer.ts:112]: Declare this long localize statement as a const.

File: src/ecs/commands/runCommandInContainer.ts [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        '--task',
+        task,
+        '--container',
+        node.continerName,
+        '--command',
+        command,
+        '--interactive'
+    ).run()
+
+    getLogger().info(response.stdout)
```
Comment (bryceitoc9) [inline src/ecs/commands/runCommandInContainer.ts:135]: Why is the output being logged? Should this be written to the output channel instead?

Comment (gsweetwood) [reply src/ecs/commands/runCommandInContainer.ts:135]: Good catch, a leftover from prototyping.  Switched to output channel.

File: src/ecs/commands/runCommandInContainer.ts (hunk 7 above)
Comment (gsweetwood) [reply src/ecs/commands/runCommandInContainer.ts:58]: done

File: src/ecs/commands/runCommandInContainer.ts (hunk 6 above)
Comment (gsweetwood) [reply src/ecs/commands/runCommandInContainer.ts:38]: Moved to a function that will check both and used the suggested call to replace 'AWS'.  Left the strings separate for clarity to the user.

File: package.nls.json (hunk 5 above)
Comment (gsweetwood) [reply package.nls.json:176]: added suggestion

File: package.nls.json (hunk 4 above)
Comment (gsweetwood) [reply package.nls.json:172]: Combined to a more generic name and replaced all instances that need it.

File: src/awsexplorer/activation.ts [hunk 10]
```diff
@@ -179,6 +179,7 @@ async function registerAwsExplorerCommands(
 
     context.subscriptions.push(
         vscode.commands.registerCommand('aws.loadMoreChildren', async (node: AWSTreeNodeBase & LoadMoreNode) => {
+            node.setPersistChildren()
             await loadMoreChildrenCommand(node, awsExplorer)
```
Comment (justinmk3) [inline src/awsexplorer/activation.ts]: why wouldn't the call go in `loadMoreChildrenCommand` itself?

File: src/ecs/explorer/ecsClusterNode.ts [hunk 11]
```diff
@@ -23,6 +23,7 @@ import { getLogger } from '../../shared/logger'
  */
 export class EcsClusterNode extends AWSTreeNodeBase implements AWSResourceNode, LoadMoreNode {
     private readonly childLoader: ChildNodeLoader
+    public persistChildren: boolean = false
```
Comment (justinmk3) [inline src/ecs/explorer/ecsClusterNode.ts]: this code is copy-pasted in 5+ classes. why not put it in `AWSTreeNodeBase` instead?

File: src/shared/treeview/nodes/loadMoreNode.ts [hunk 12]
```diff
@@ -45,4 +45,15 @@ export interface LoadMoreNode {
...
+    /**
+     * Sets persistChildren value to true
+     */
+    setPersistChildren(): void
+
+    /**
+     * Instructs the node to persist the child nodes during getChildren(). A false value will clear the children nodes
+     * during getChildren()
+     */
+    persistChildren: boolean
```
Comment (justinmk3) [inline src/shared/treeview/nodes/loadMoreNode.ts]: this is mutable so why is setPersistChildren() needed?

File: src/s3/explorer/s3FolderNode.ts [hunk 13]
```diff
@@ -47,6 +48,11 @@ export class S3FolderNode extends AWSTreeNodeBase implements AWSResourceNode, Lo
     }
 
     public async getChildren(): Promise<AWSTreeNodeBase[]> {
+        if (!this.persistChildren) {
+            this.clearChildren()
+        } else {
+            this.persistChildren = false
+        }
```
Comment (justinmk3) [inline src/s3/explorer/s3FolderNode.ts]: Why wasn't this needed on s3FolderNode, until now?

File: src/ecs/explorer/ecsNode.ts [hunk 14]
```diff
@@ -17,14 +17,21 @@ import { getLogger } from '../../shared/logger/logger'
...
     public constructor(private readonly ecs: EcsClient) {
         super('ECS', vscode.TreeItemCollapsibleState.Collapsed)
         this.childLoader = new ChildNodeLoader(this, token => this.loadPage(token))
-        this.update()
+        this.contextValue = 'awsEcsNode'
     }
 
     public async getChildren(): Promise<AWSTreeNodeBase[]> {
+        // This helps decipher when this method is called by a loadMoreChildren command or an explorer refresh
+        if (!this.persistChildren) {
```
[the remark this answers is not in the document]
Comment (justinmk3) [reply src/ecs/explorer/ecsNode.ts]: Fixed by https://github.com/aws/aws-toolkit-vscode/pull/1921

File: package.nls.json [hunk 15]
```diff
@@ -29,6 +29,7 @@
     "AWS.codelens.lambda.configEditor": "Edit Debug Configuration (Beta)",
     "AWS.configuration.title": "AWS Configuration",
     "AWS.configuration.profileDescription": "The name of the credential profile to obtain credentials from.",
+    "AWS.configuration.description.ecs.warnRunCommand": "Prompt the user before executing a command in a running ECS container.",
```
Comment (justinmk3) [inline package.nls.json:32]: we don't usually refer to the user in the third person
```suggestion
    "AWS.configuration.description.ecs.warnRunCommand": "Prompt before executing a command in an ECS container.",
```

File: src/ecs/commands/runCommandInContainer.ts [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+                node.continerName
+            ),
+            { modal: true },
+            localize('AWS.generic.response.yes', 'Yes'),
+            localize('AWS.message.prompt.yesDontAskAgain', "Yes, and don't ask again")
+        )
+        if (choice === undefined) {
+            getLogger().debug('ecs: Cancelled runCommandInContainer')
+            return
+        } else if (choice === localize('AWS.message.prompt.yesDontAskAgain', "Yes, and don't ask again")) {
```
Comment (justinmk3) [inline src/ecs/commands/runCommandInContainer.ts:109]: could have a local var instead of repeating this localize() calculation

File: src/ecs/commands/runCommandInContainer.ts [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+        'aws',
+        undefined,
+        'ecs',
+        'execute-command',
+        '--cluster',
+        node.clusterArn,
+        '--task',
+        task,
+        '--container',
+        node.continerName,
```
Comment (justinmk3) [inline src/ecs/commands/runCommandInContainer.ts:125]: typo

File: src/ecs/commands/runCommandInContainer.ts [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+        recordEcsRunExecuteCommand({ result: 'Succeeded', ecsExecuteCommandType: 'command' })
+    }
+}
+
+async function verifyCliAndPlugin(window: Window): Promise<void> {
+    const verifyAwsCliResponse = await new ChildProcess(true, 'aws', undefined, '--version').run()
+    if (verifyAwsCliResponse.exitCode !== 0) {
+        const noCli = localize(
+            'AWS.command.ecs.runCommandInContainer.noCliFound',
+            'Please install the {0} CLI before proceding',
```
Comment (justinmk3) [inline src/ecs/commands/runCommandInContainer.ts:145]:
```suggestion
            'Install the {0} CLI before proceeding',
```

File: src/ecs/commands/runCommandInContainer.ts [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+        window.showErrorMessage(noCli)
+        throw new Error('The AWS CLI is not installed.')
+    }
+
+    const verifySsmPluginResponse = await new ChildProcess(true, 'session-manager-plugin').run()
+    if (verifySsmPluginResponse.exitCode !== 0) {
+        window.showErrorMessage(
+            localize(
+                'AWS.command.ecs.runCommandInContainer.noPluginFound',
+                'Please install the SSM Session Manager plugin before proceding'
```
Comment (justinmk3) [inline src/ecs/commands/runCommandInContainer.ts:157]:
```suggestion
                'Install the SSM Session Manager plugin before proceding'
```

File: src/ecs/commands/runCommandInContainer.ts [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+            )
+        )
+        throw new Error('The SSM Session Manager plugin is not installed.')
```
Comment (justinmk3) [inline src/ecs/commands/runCommandInContainer.ts:160]: Why throw an error if we are using showErrorMessage()? Does this manifest as multiple messages?

File: src/shared/clients/ecsClient.ts [hunk 21]
```diff
@@ -77,6 +77,34 @@ export class DefaultEcsClient {
         }
     }
 
+    public async listTasks(cluster: string, serviceName: string): Promise<string[]> {
+        const sdkClient = await this.createSdkClient()
+
+        const params: ECS.ListTasksRequest = { cluster: cluster, serviceName: serviceName, desiredStatus: 'RUNNING' }
```
Comment (justinmk3) [inline src/shared/clients/ecsClient.ts:83]: should the docstring mention that the function only gets RUNNING tasks?

File: src/ecs/commands/runCommandInContainer.ts [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+        task,
+        '--container',
+        node.continerName,
+        '--command',
+        command,
+        '--interactive'
+    ).run()
+
+    if (response.exitCode !== 0) {
+        getLogger().error(response.stderr)
```
Comment (justinmk3) [inline src/ecs/commands/runCommandInContainer.ts:132]: - need some sort of prefix, otherwise this is unreadable in the logs. 
- storing ChildProcess as a local var allows us to get the string representation of it, for logging.

```suggestion
        const cmd = await new ChildProcess(...)
        const response = cmd.run()
        getLogger().error('ECS command failed: %O: %O', cmd, response.stderr)
```

File: src/ecs/commands/runCommandInContainer.ts [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+        return {
+            label: task.taskArn!.substr(-32),
+            detail: `Status: ${task.lastStatus}  Desired status: ${task.desiredStatus}`,
+        }
+    })
+    if (quickPickItems.length === 0) {
+        window.showInformationMessage(
+            localize(
+                'AWS.command.ecs.runCommandInContainer.noTasks',
+                'There are no running tasks for the service: {0}',
```
Comment (justinmk3) [inline src/ecs/commands/runCommandInContainer.ts:43]:
```suggestion
                'No running tasks for service: {0}',
```

File: src/ecs/commands/runCommandInContainer.ts (hunk 17 above)
Comment (gsweetwood) [reply src/ecs/commands/runCommandInContainer.ts:142]: good catch, renamed at the node level

File: src/ecs/commands/runCommandInContainer.ts (hunk 20 above)
Comment (gsweetwood) [reply src/ecs/commands/runCommandInContainer.ts:178]: Error not really needed here.  Changed so that `runCommandInContainer `will simply return after displaying the message.

File: src/ecs/commands/runCommandInContainer.ts (hunk 22 above)
Comment (gsweetwood) [reply src/ecs/commands/runCommandInContainer.ts:150]: That's really useful, thanks.

File: src/shared/clients/ecsClient.ts (hunk 21 above)
Comment (gsweetwood) [reply src/shared/clients/ecsClient.ts:81]: I think I will be changing this to return all tasks (same with decribeTasks) and do any filtering at the function level.

File: package.nls.json (hunk 15 above)
Comment (gsweetwood) [reply package.nls.json]: Phrasing was changed to your suggestion.

File: src/ecs/commands/runCommandInContainer.ts [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+import { ChildProcess } from '../../shared/utilities/childProcess'
+import { EcsContainerNode } from '../explorer/ecsContainerNode'
+import { recordEcsRunExecuteCommand } from '../../shared/telemetry/telemetry.gen'
+import { DefaultSettingsConfiguration } from '../../shared/settingsConfiguration'
+import { extensionSettingsPrefix } from '../../shared/constants'
+import { getIdeProperties } from '../../shared/extensionUtilities'
+import { showOutputMessage } from '../../shared/utilities/messages'
+import { ext } from '../../shared/extensionGlobals'
+
+export async function runCommandInContainer(
```
Comment (JadenSimon) [inline src/ecs/commands/runCommandInContainer.ts:21]: Can we break this function up? It's doing a lot of things with many early returns.

File: src/shared/clients/ecsClient.ts [hunk 25]
```diff
@@ -77,6 +75,39 @@ export class DefaultEcsClient {
...
 
+    public async listTasks(cluster: string, serviceName: string): Promise<string[]> {
+        const sdkClient = await this.createSdkClient()
+
+        const params: ECS.ListTasksRequest = { cluster: cluster, serviceName: serviceName }
+        try {
+            const listTasksResponse = await sdkClient.listTasks(params).promise()
+            return listTasksResponse.taskArns ?? []
+        } catch (error) {
+            getLogger().error(
```
Comment (JadenSimon) [inline src/shared/clients/ecsClient.ts:86]: Just let the errors bubble up and handle them further up the stack.

File: src/ecs/commands/runCommandInContainer.ts [hunk 26]
```diff
@@ -0,0 +1,189 @@
...
+    outputChannel = ext.outputChannel
+): Promise<void> {
+    getLogger().debug('RunCommandInContainer called for: %O', node.containerName)
+
+    if (!(await verifyCliAndPlugin(window))) {
+        return
+    }
+
+    // Check to see if there are any deployments in process
+    const deployments = (await node.ecs.describeServices(node.clusterArn, [node.serviceName]))[0].deployments
```
Comment (JadenSimon) [inline src/ecs/commands/runCommandInContainer.ts:33]: We're not catching any of the errors that these client calls can throw.

File: src/shared/clients/ecsClient.ts [hunk 27]
```diff
@@ -77,6 +75,39 @@ export class DefaultEcsClient {
...
+        try {
+            const describedTasks = await sdkClient.describeTasks(params).promise()
+            return describedTasks.tasks ?? []
+        } catch (error) {
+            getLogger().error(error)
+            throw error
+        }
+    }
+
+    public async describeServices(cluster: string, services: string[]): Promise<ECS.Service[]> {
```
Comment (JadenSimon) [inline src/shared/clients/ecsClient.ts:106]: `listServices` is now calling `describeServices` too. So either revert the change and have `listServices` call `listServices` or remove this method and rename the other one to `describeServices`

Comment (gsweetwood) [reply src/shared/clients/ecsClient.ts:106]: There are two API calls needed, `listServices` and `describeServices`.  To properly display the info we want, both calls are needed. We also need `describeServices` on it's own for other commands.  I have changed the name of any functions that utilize more than one API call as `get<Resource>` to try and show it's not just simply the return of a single API call.
